### PR TITLE
Add configuration name sanitization to Jucer2Reproducer

### PR DESCRIPTION
**Pull request by TheSlowGrowth**
`Jucer2Reproducer` now removes invalid characters from configuration names and ensures that they are unique to each exporter by adding digits if required. If the name was changed, a warning is printed to cerr.

### CODE REVIEW

File: Jucer2Reprojucer/main.cpp [hunk 1]
```diff
@@ -270,6 +270,51 @@ void writeUserNotes(LineWriter& wLn, const juce::ValueTree& valueTree)
...
+            }
+            else if (originalName != sanitizedName)
+                std::cerr << "warning: Configuration '" << originalName << "' in exporter '" << exporterName
+                          << "' contains invalid characters. The name was changed to '" << sanitizedName << "'" << std::endl;
+
+            // add digits until unique
+            int number = 1;
+            auto sanitizedNameWithOptionalNumber = sanitizedName;
+            while (sanitizedConfigNames.contains(sanitizedNameWithOptionalNumber))
+                sanitizedNameWithOptionalNumber = sanitizedName + juce::String(number++);
```
Comment (McMartin) [inline Jucer2Reprojucer/main.cpp]: This is a great idea! However, there is an edge case that this implementation doesn't cover: when exporters don't have the configurations in the same order, or when some configurations only exist on certain exporters.

For instance:
```yaml
- XCODE_MAC:
  - configurations:
    - name: "Config (-feature)"
      defines: "FEATURE=0"
    - name: "Config (+feature)"
      defines: "FEATURE=1"

- LINUX_MAKE:
  - configurations:
    - name: "Config (+feature)"
      defines: "FEATURE=1"
    - name: "Config (-feature)"
      defines: "FEATURE=0"
```
would be converted to 
```yaml
- XCODE_MAC:
  - configurations:
    - name: "Configfeature"
      defines: "FEATURE=0"
    - name: "Configfeature1"
      defines: "FEATURE=1"

- LINUX_MAKE:
  - configurations:
    - name: "Configfeature"
      defines: "FEATURE=1"
    - name: "Configfeature1"
      defines: "FEATURE=0"
```
which would break the matching of the configuration names between the exporters.

To solve this issue, we should collect all configuration names, make a mapping of original to sanitized configuration names, and then use that mapping when writing the configuration names to the `CMakeLists.txt` file.

File: Jucer2Reprojucer/main.cpp [hunk 2]
```diff
@@ -270,6 +270,51 @@ void writeUserNotes(LineWriter& wLn, const juce::ValueTree& valueTree)
...
+
+        auto exporter = exporterSettingsWithInvalidConfigNames.createCopy();
+        juce::StringArray sanitizedConfigNames;
+
+        for (int i = 0; i < exporter.getNumChildren(); i++)
+        {
+            // remove invalid characters
+            auto configuration = exporter.getChild(i);
+            auto originalName = configuration.getProperty("name").toString();
+            auto sanitizedName = originalName.retainCharacters(validCharacters);
```
Comment (McMartin) [inline Jucer2Reprojucer/main.cpp]: Removing invalid characters does the job, but it's not very nice for configuration names like `Debug (all warnings)` or `Release (with LTO)`. 

Do you think you could come up with an algorithm that:
- replace invalid characters with `_`,
- keep at most one consecutive replacement `_` (i.e. it's fine to have several consecutive `_` characters if they were already in the original configuration name),
- don't add a replacement `_` at the end?

That would make `Debug (all warnings)` -> `Debug_all_warnings` and `Release (with LTO)` -> `Release_with_LTO`

File: Jucer2Reprojucer/main.cpp [hunk 3]
```diff
@@ -270,6 +270,51 @@ void writeUserNotes(LineWriter& wLn, const juce::ValueTree& valueTree)
   }
 }
 
+    juce::ValueTree sanitizeConfigurationNames(const juce::String& exporterName, const juce::ValueTree& exporterSettingsWithInvalidConfigNames)
```
Comment (McMartin) [inline Jucer2Reprojucer/main.cpp]: I know it is not written down explicitly, and there is no CI check for it, but the C++ code of FRUT is formatted using `clang-format`. There is a [`.clang-format` ](https://github.com/McMartin/FRUT/blob/c3dd1775bba0ad0db43389203658151e4db77198/.clang-format) file at the root of the repository, which is currently configured to be used with `clang-format-7.0.0` (you can download binaries for most platforms on http://releases.llvm.org/download.html#7.0.0). I usually run the following:
```
clang-format-7.0.0 -i -style=file -verbose Jucer2Reprojucer/main.cpp
```

File: Jucer2Reprojucer/main.cpp (hunk 2 above)
Comment (TheSlowGrowth) [reply Jucer2Reprojucer/main.cpp]: Sure, good idea

File: Jucer2Reprojucer/main.cpp (hunk 1 above)
Comment (TheSlowGrowth) [reply Jucer2Reprojucer/main.cpp]: Hmm, I see. Yes, the matching would break. You're right, that should be avoided. I'll make the required changes.

File: Jucer2Reprojucer/main.cpp (hunk 2 above)
Comment (TheSlowGrowth) [reply Jucer2Reprojucer/main.cpp]: `makeValidIdentifier` seems to do most of this already, except than only some characters are replaced with `_`, other are simply removed (this includes the space).

Now that we're dealing with multiple name-sanitization functions, it would be good to a) merge them into one or b) alter their names to be more clear about their use.
Forthe sake of readability and simplicity I would prefer to use `makeValidIdentifier` for the config name sanitization as well and add the required changes to this function instead of writing another one.

Do you agree?

File: Jucer2Reprojucer/main.cpp (hunk 3 above)
Comment (TheSlowGrowth) [reply Jucer2Reprojucer/main.cpp]: Yes, sure I can run clang-format on it. I simply didn't notice it. Thanks for pointing it out.

I would suggest to add a check to the CI. [run-clang-format](https://github.com/Sarcasm/run-clang-format) could help with this.

File: Jucer2Reprojucer/main.cpp (hunk 2 above)
Comment (McMartin) [reply Jucer2Reprojucer/main.cpp]: Jucer2Reprojucer's `makeValidIdentifier` is a stripped-down version of [a function named the same in Projucer](https://github.com/WeAreROLI/JUCE/blob/5.4.4/extras/Projucer/Source/Utility/Helpers/jucer_CodeHelpers.cpp#L71-L123). It is then used in Jucer2Reprojucer exactly like in Projucer. We shouldn't change the logic of this function, otherwise Jucer2Reprojucer will diverge from Projucer.  I agree that there should be some documentation (or at least a comment) on the function to explain that.

We could indeed use `makeValidIdentifier` to sanitize configuration names, but it wouldn't be as nice as the algorithm I proposed (`Debug (all warnings)` would become `Debugallwarnings`). 

If you are concerned about code reuse, don't worry, this new function to convert a string into a (nice looking) valid CMake identifier will be perfect to compute `escapedJucerFileName` as well:
https://github.com/McMartin/FRUT/blob/dfd36f97e0c654f379d3f7a6034d50be2efef326/Jucer2Reprojucer/main.cpp#L546-L552

File: Jucer2Reprojucer/main.cpp (hunk 3 above)
Comment (McMartin) [reply Jucer2Reprojucer/main.cpp]: > I would suggest to add a check to the CI.

I'll work on that when I have some time 😅

Comment (McMartin) [reply Jucer2Reprojucer/main.cpp]: Done in https://github.com/McMartin/FRUT/pull/526.

**Comment by TheSlowGrowth** [comment]
@McMartin Does this look good to you? Still some cleanup and squashing left to do but the direction should hopefully be correct. I'll continue tomorrow.
Side note: clang-format messes with the comments. Could comment reflowing be switched off? Otherweise I'll exclude the comments from clang-format, if requried.

**Comment by TheSlowGrowth** [comment]
@McMartin please review

File: Jucer2Reprojucer/main.cpp [hunk 4]
```diff
@@ -270,6 +270,106 @@ void writeUserNotes(LineWriter& wLn, const juce::ValueTree& valueTree)
...
+}
+
+// clang-format off
+/** Reads a jucerProject and sanitizes configuration names. Valid config names match the
+  * regex ^[A-Za-z0-9_]*$. Sanitization is done with sanitizeName(). An optional digit is 
+  * added until the configuration name is unique. The sanitization is synchronized across 
+  * all exporters in the project.
+  */
+// clang-format on
+void sanitizeConfigurationNames(const juce::ValueTree& jucerProject)
```
Comment (McMartin) [inline Jucer2Reprojucer/main.cpp]: This function is pretty well written, but it doesn't fit the architecture of `Jucer2Reprojucer`. More concretely, `Jucer2Reprojucer`:
- never modifies the original `ValueTree`,
- extracts data from the `ValueTree` right before transforming it and writing it to the `CMakeLists.txt` file, and
- tries to read the content of the `ValueTree` only once.

Could you please try to take the content of this function and move it to where we are already looping over `EXPORTFORMATS` and `CONFIGURATIONS`? This would guarantee that we are iterating in the same way when sanitizing and when outputting to the `CMakeLists.txt` file. It would also allow us to write the original configuration name as a comment when the sanitized name is different:
```cmake
jucer_export_target_configuration(
  "Visual Studio 2017"
  NAME "Debug_32_bit"  # "Debug (32-bit)" in MyGreatProject.jucer
  DEBUG_MODE ON
)
```
Finally, the warnings would be properly interleaved with other warnings that we might add in the future.

File: Jucer2Reprojucer/main.cpp [hunk 5]
```diff
@@ -543,14 +645,7 @@ int main(int argc, char* argv[])
...
 
-  auto escapedJucerFileName = jucerFileName.toStdString();
-  std::replace_if(
-    escapedJucerFileName.begin(), escapedJucerFileName.end(),
-    [](const std::string::value_type& c) {
-      return !(std::isalpha(c, std::locale::classic())
-               || std::isdigit(c, std::locale::classic()));
-    },
-    '_');
+  auto escapedJucerFileName = sanitizeName(jucerFileName);
```
Comment (McMartin) [inline Jucer2Reprojucer/main.cpp]: Please make this cleanup in its own commit. I think there should be 3 commits in this PR:
- one for introducing the `sanitizeName` function,
- one for using the `sanitizeName` function on configuration names, and
- one for using the `sanitizeName` function here.

**Comment by McMartin** [comment]
@TheSlowGrowth Happy new year! Thanks a lot for all the good work on this PR. I would like to see it merged in the coming month. Do you intend to continue working on it or can I take over?

**Comment by McMartin** [comment]
@TheSlowGrowth I would like to see this PR merged and #520 closed in a week or two. Do you have time to work on this topic or should I handle it myself? If I don't hear from you by February 8th, 2020, I'll  rebase. rework and merge this PR on my own.

**Comment by TheSlowGrowth** [comment]
@McMartin I wasn't logged into github for quite some time (RL ...) and just saw your messages. Please excuse that I didn't see this earlier.
I have just pushed a new attempt. Please take a look if this fixes all issues. I have tested it myself and it seems fine, though I'm afraid I might have missed some spots where the sanitization would have to be added.

**Comment by McMartin** [comment]
@TheSlowGrowth no problem at all, you're entirely excused.

I'll do a quick review now and see how we should handle clang. Maybe silencing new warnings is fine.

File: Jucer2Reprojucer/main.cpp [hunk 6]
```diff
@@ -151,6 +151,81 @@ void printError(const juce::String& error)
...
+  if (sanitizedConfigNameTable.find(originalName) != sanitizedConfigNameTable.end())
+    return sanitizedConfigNameTable[originalName];
+
+  // need to sanitize
+  sanitizedName = sanitizeName(originalName);
+
+  // add digits until unique
+  int number = 1;
+  auto sanitizedNameWithoutNumber = sanitizedName;
+  while (sanitizedConfigNameTable.find(sanitizedName) != sanitizedConfigNameTable.end())
```
Comment (McMartin) [inline Jucer2Reprojucer/main.cpp]: `sanitizedConfigNameTable.find(sanitizedName)` will always return `sanitizedConfigNameTable.end()` because `sanitizedName` is only used as a value of the map, not as a key.
Using 2 `juce::StringArray` objects, one for the original names and one for the sanitized names, might result in simpler code.

File: Jucer2Reprojucer/main.cpp [hunk 7]
```diff
@@ -151,6 +151,82 @@ void printError(const juce::String& error)
...
+      output += *c;
+    }
+    // replace invalid characters with "_"
+    else
+    {
+      // don't add "_" at the beginning
+      if (output.isNotEmpty())
+        mustInsertReplacementChar = true;
+    }
+    c++;
```
Comment (McMartin) [inline Jucer2Reprojucer/main.cpp]: This happens on every iteration of the loop, so it should be in the third clause of the `for` statement.

File: Jucer2Reprojucer/main.cpp [hunk 8]
```diff
@@ -151,6 +151,82 @@ void printError(const juce::String& error)
...
+ */
+juce::String sanitizeConfigurationName(const juce::String& originalName)
+{
+  static std::map<juce::String, juce::String> sanitizedNameTable;
+  static juce::StringArray sanitizedNameList;
+  juce::String sanitizedName;
+
+  // check if this name was sanitized before
+  if (sanitizedNameTable.find(originalName) != sanitizedNameTable.end())
+    return sanitizedNameTable[originalName];
```
Comment (McMartin) [inline Jucer2Reprojucer/main.cpp]: If `originalName` was sanitized before, the function will return here without having used `sanitizedName` at all. Please declare and initialize `sanitizedName` only when you need it.

File: Jucer2Reprojucer/main.cpp [hunk 9]
```diff
@@ -151,6 +151,82 @@ void printError(const juce::String& error)
...
+
+/** Sanitizes a configuration name. Valid config names match the regex ^[A-Za-z0-9_]*$.
+ *  Sanitization is done with sanitizeName(). An optional digit is added until the
+ *  configuration name is unique. The sanitization is synchronized across
+ *  the entire project.
+ */
+juce::String sanitizeConfigurationName(const juce::String& originalName)
+{
+  static std::map<juce::String, juce::String> sanitizedNameTable;
+  static juce::StringArray sanitizedNameList;
```
Comment (McMartin) [inline Jucer2Reprojucer/main.cpp]: Similarly to `sanitizedName`, please declare and initialize `sanitizedNameList` only when you need it. Even if it is a `static` variable, it doesn't need to be at the top of the function.

File: Jucer2Reprojucer/main.cpp [hunk 10]
```diff
@@ -2006,9 +2006,19 @@ int main(int argc, char* argv[])
...
         wLn("  \"", exporterName, "\"");
-        wLn("  NAME \"", configuration.getProperty("name").toString(), "\"");
+        if (originalConfigName != sanitizedConfigName)
+        {
+          const auto jucerFileBasename = jucerFile.getFileName();
+          wLn("  NAME \"", sanitizedConfigName, "\" # \"", originalConfigName, "\" in ",
+              jucerFileBasename);
+        }
+        else
+          wLn("  NAME \"", sanitizedConfigName, "\"");
```
Comment (McMartin) [inline Jucer2Reprojucer/main.cpp]: Please put curly braces around this statement.

File: Jucer2Reprojucer/CMakeLists.txt [hunk 11]
```diff
@@ -103,7 +103,7 @@ jucer_project_module(
 
 jucer_export_target(
   "Xcode (MacOSX)"
-  COMPILER_FLAGS_FOR_maximum_warnings "-Werror -Weverything -Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-padded"
+  COMPILER_FLAGS_FOR_maximum_warnings "-Werror -Weverything -Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-exit-time-destructors -Wno-padded"
```
Comment (McMartin) [inline Jucer2Reprojucer/CMakeLists.txt]: Adding `-Wno-exit-time-destructors` to the compiler flags of `Jucer2Reprojucer` should be done in the same commit that introduces the static variables, otherwise that commit won't build.

**Comment by McMartin** [comment]
Superseded by https://github.com/McMartin/FRUT/pull/655.
